networkx/utils/backends.py

```diff
@@ -0,0 +1,654 @@
+
+__all__ = ["_dispatch", "_mark_tests"]
+
+
+class PluginInfo:
+    """Lazily loaded entry_points plugin information"""
+
+    def __init__(self):
+        self._items = None
+
+    def __bool__(self):
+        return len(self.items) > 0
+
+    @property
+    def items(self):
+        if self._items is None:
+            if sys.version_info < (3, 10):
+                self._items = entry_points()["networkx.plugins"]
+            else:
+                self._items = entry_points(group="networkx.plugins")
+        return self._items
+
+    def __contains__(self, name):
+        if sys.version_info < (3, 10):
+            return len([ep for ep in self.items if ep.name == name]) > 0
+        return name in self.items.names
+
+    def __getitem__(self, name):
+        if sys.version_info < (3, 10):
+            return [ep for ep in self.items if ep.name == name][0]
+        return self.items[name]
+
+
+plugins = PluginInfo()
+_registered_algorithms = {}
+
+
+def _register_algo(name, wrapped_func):
+    if name in _registered_algorithms:
+        raise KeyError(f"Algorithm already exists in dispatch registry: {name}")
+    _registered_algorithms[name] = wrapped_func
+    wrapped_func.dispatchname = name
+
+
+def _dispatch(
+    func=None,
+    *,
+    name=None,
+    graphs="G",
+    edge_attrs=None,
+    node_attrs=None,
+    preserve_edge_attrs=False,
+    preserve_node_attrs=False,
+    preserve_graph_attrs=False,
+):
+    """Dispatches to a backend algorithm based on input graph types.
+
+    Parameters
+    ----------
+    func : function
+
+    name : str, optional
+        The name of the algorithm to use for dispatching. If not provided,
+        the name of ``func`` will be used. ``name`` is useful to avoid name
+        conflicts, as all dispatched algorithms live in a single namespace.
+
+    graphs : str or dict, default "G"
+        If a string, the parameter name of the graph, which must be the first
+        argument of the wrapped function. If more than one graph is required
+        for the algorithm (or if the graph is not the first argument), provide
+        a dict of parameter name to argument position for each graph argument.
+        For example, ``@_dispatch(graphs={"G": 0, "auxiliary?": 4})``
+        indicates the 0th parameter ``G`` of the function is a required graph,
+        and the 4th parameter ``auxiliary`` is an optional graph.
+        To indicate an argument is a list of graphs, do e.g. ``"[graphs]"``.
+
+    edge_attrs : str or dict, optional
+        ``edge_attrs`` may be strings of the function argument that indicate
+        a single edge attribute to use to convert the graph. They may also be
+        dicts that map argument name of attributes to argument name of default
+        values. The default value may be a non-string value such as 0.
+        If not provided, the default value is assumed to be 1.
+        To indicate an argument is a list of attributes, do e.g. ``"[attrs]"``.
```

Comment:
> This doesn't handle the case where a default value is a string.

To clarify, a user can still pass a string as the default value, and this will work. We don't yet support a string as the default default value, which I don't think exists in the networkx API.